### PR TITLE
Test: markup renderer

### PR DESCRIPTION
**Pull request by toaster**
This PR is my proposal on how to test rendered output without images.
It introduces a markup renderer which creates a markup representation of the render tree.
Common icons and colours are displayed by name instead by filename or colour values.

This is a draft for now because there are a lot of tests that still need to be converted and maybe there might be adjustments to the markup renderer during this process.

### CODE REVIEW

**Comment by andydotxyz** [comment]
This is totally cool :).
I wonder though if it could be done without the `heredoc.Doc` dependency?
If we could just pass raw strings it would be much neater, perhaps this calls for another test helper?

**Review by benet37 — APPROVED** [review]

**Comment by toaster** [comment]
> I wonder though if it could be done without the `heredoc.Doc` dependency?
> If we could just pass raw strings it would be much neater, perhaps this calls for another test helper?

It should be possible. You’re thinking of a helper along the lines:
```go
func AssertMarkup(t *testing.T, expected string, actual string)
```
or even:
```go
func AssertMarkup(t *testing.T, expected string, c fyne.Canvas)
```

**Comment by andydotxyz** [comment]
> It should be possible. You’re thinking of a helper along the lines:

Yeah looks good

**Comment by toaster** [comment]
This PR is ready to land, I think.
It contains a markup renderer for tests and an assertion method to access it.
I switched all public widget specs and some other to markup based instead of image based.

**Comment by toaster** [comment]
This PR now included the (stripped down version of) #1403.
This was needed to ensure distinguishable colours, which became necessary because I had to switch to map based lookup to reduce cyclomatic complexity.

File: test/testtheme.go [hunk 1]
```diff
@@ -8,7 +8,7 @@ import (
 )
 
 var (
-	red   = &color.RGBA{R: 255, G: 0, B: 0, A: 255}
```
Comment (andydotxyz) [inline test/testtheme.go:11]: Is this change required? I'm not sure what it imacts

File: test/theme.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+// Theme returns a theme useful for image based tests.
+func Theme() fyne.Theme {
+	if defaulTheme == nil {
+		defaulTheme = &configurableTheme{
+			background:         color.NRGBA{R: 0x44, G: 0x44, B: 0x44, A: 0xff},
+			bold:               bold,
+			boldItalic:         bolditalic,
+			button:             color.NRGBA{R: 0x33, G: 0x33, B: 0x33, A: 0xff},
+			disabledButton:     color.NRGBA{R: 0x22, G: 0x22, B: 0x22, A: 0xff},
+			disabledIcon:       color.NRGBA{R: 0xff, G: 0x00, B: 0x00, A: 0xff}, // deprecated: outstanding red variant to make it visible
```
Comment (andydotxyz) [inline test/theme.go:49]: I think maybe "bright" not "outstanding"

File: test/markup_renderer.go [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+	r.writeIndent()
+	r.writeCloseTag("canvas")
+}
+
+func (r *markupRenderer) writeCanvasObject(obj fyne.CanvasObject, _, clipPos fyne.Position, clipSize fyne.Size) bool {
+	attrs := map[string]*string{}
+	r.setPosAttr(attrs, "pos", obj.Position())
+	r.setSizeAttr(attrs, "size", obj.Size())
+	if clipPos != fyne.NewPos(0, 0) || clipSize != fyne.NewSize(math.MaxInt32, math.MaxInt32) {
+		r.setRectAttr(attrs, "clip", clipPos, clipSize)
```
Comment (andydotxyz) [inline test/markup_renderer.go:229]: I don't think this should be rendered - it is a property of the driver / renderer not the widget.
Also it gets in the way of this markup being writable manually...

**Review by andydotxyz — CHANGES_REQUESTED** [review]

File: test/testtheme.go (hunk 1 above)
Comment (toaster) [reply test/testtheme.go:11]: With the former colour the background icon of dialogs was not visible at all. With this it is at least faintly visible.

**Comment by andydotxyz** [comment]
It's failing static analysis:
```
/home/runner/work/fyne/fyne/test/markup_renderer.go
  (129, 26)  U1000  func (*markupRenderer).setRectAttr is unused
```

**Review by andydotxyz — DISMISSED** [review]
This resolves my comments, thanks. I would like to get @stuartmscott feedback as well, I think he returns next week.

File: test/markup_renderer.go [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+
+func (r *markupRenderer) writeIndent() {
+	for i := 0; i < r.indentation; i++ {
+		r.w.WriteRune('\t')
+	}
+}
+
+func (r *markupRenderer) writeImage(i *canvas.Image, attrs map[string]*string) {
+	r.setStringAttr(attrs, "file", i.File)
+	r.setResourceAttr(attrs, "rsc", i.Resource)
```
Comment (stuartmscott) [inline test/markup_renderer.go:297]: It took me a while to figure this out, I think we can afford the storage space of using full names; `resource`

File: test/markup_renderer.go [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+	r.setBoolAttr(attrs, "img", i.Image != nil)
```
Comment (stuartmscott) [inline test/markup_renderer.go:298]: Ditto; `image`

File: test/markup_renderer.go [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+		r.writeCloseTag("overlay")
+	}
+	r.indentation--
+	r.writeIndent()
+	r.writeCloseTag("canvas")
+}
+
+func (r *markupRenderer) writeCanvasObject(obj fyne.CanvasObject, _, _ fyne.Position, _ fyne.Size) bool {
+	attrs := map[string]*string{}
+	r.setPosAttr(attrs, "pos", obj.Position())
```
Comment (stuartmscott) [inline test/markup_renderer.go:225]: Ditto; `position`

File: dialog/base_test.go [hunk 7]
```diff
@@ -7,23 +7,24 @@ import (
...
 
 	d := NewCustom("Title", "OK", label, w)
 
 	d.Show()
-	test.AssertImageMatches(t, "dialog-custom-dark.png", w.Canvas().Capture())
+	test.AssertImageMatches(t, "dialog-custom-default.png", w.Canvas().Capture())
 
-	test.ApplyTheme(t, theme.LightTheme())
-	test.AssertImageMatches(t, "dialog-custom-light.png", w.Canvas().Capture())
+	test.ApplyTheme(t, test.NewTheme())
```
Comment (stuartmscott) [inline dialog/base_test.go:28]: Maybe we should go all in and call this `test.UglyTheme()`

**Review by stuartmscott — COMMENTED** [review]
Few naming nits, otherwise looks good!

Have you considered how this can be used in the reverse? This would allow developers to define a UI in markup and then convert it into a canvas object. The main challenges I foresee here are mapping resource name to bundled resource, and defining the callbacks eg button.OnTapped, list.Length, list.CreateItem, et al.

File: dialog/base_test.go (hunk 7 above)
Comment (toaster) [reply dialog/base_test.go:28]: > Have you considered how this can be used in the reverse? This would allow developers to define a UI in markup and then convert it into a canvas object. The main challenges I foresee here are mapping resource name to bundled resource, and defining the callbacks eg button.OnTapped, list.Length, list.CreateItem, et al.

No, I didn’t. And, honestly, I am not a fan of defining UI via markup. I prefer to write the code.

**Comment by toaster** [comment]
> Have you considered how this can be used in the reverse? This would allow developers to define a UI in markup and then convert it into a canvas object. The main challenges I foresee here are mapping resource name to bundled resource, and defining the callbacks eg button.OnTapped, list.Length, list.CreateItem, et al.

No, I didn’t. And, honestly, I am not a fan of defining UI via markup. I prefer to write the code.

File: dialog/base_test.go (hunk 7 above)
Comment (toaster) [reply dialog/base_test.go:28]: We might, but not in this PR :).

File: test/markup_renderer.go (hunk 4 above)
Comment (andydotxyz) [reply test/markup_renderer.go:292]: I think the #fyne-votes channel is showing that "res" might be more liked

File: test/markup_renderer.go (hunk 6 above)
Comment (andydotxyz) [reply test/markup_renderer.go:220]: We use "pos" already - i.e. fyne.NewPos

File: test/markup_renderer.go (hunk 5 above)
Comment (andydotxyz) [reply test/markup_renderer.go:293]: I think that img is a pretty reasonable abbreviation. It only removes 2 letters and "file" is 4 so I don't have strong feelings either way.

Comment (toaster) [reply test/markup_renderer.go:293]: The vote has clearly favoured the abbreviations (no one voted for full names).

File: test/markup_renderer.go (hunk 6 above)
Comment (toaster) [reply test/markup_renderer.go:220]: The vote has clearly favoured the abbreviations (no one voted for full names).

File: test/markup_renderer.go (hunk 4 above)
Comment (toaster) [reply test/markup_renderer.go:292]: `rsc` and `res` both got 2 votes. Therefore, I won’t change this 😜.

File: test/test.go [hunk 8]
```diff
@@ -32,6 +34,18 @@ func AssertImageMatches(t *testing.T, masterFilename string, img image.Image, ms
...
 
+// AssertRendersToMarkup asserts that the given canvas renders the expected markup.
+// The expected markup is stripped by leading common whitespace retaining the relative indentation
+// and thus matching the indentation of the output of the used markup renderer.
+//
+// Be aware, that the indentation has to use tab characters ('\t') instead of spaces.
+// Every element starts on a new line indented one more than its parent.
+// Closing elements stand on their own line, too, using the same indentation as the opening element.
+// The only exception to this are text elements which do not contain line breaks unless the text includes them.
+func AssertRendersToMarkup(t *testing.T, expected string, c fyne.Canvas, msgAndArgs ...interface{}) bool {
```
Comment (andydotxyz) [inline test/test.go:45]: Can you please add
```
//
// Since: 2.0.0
```

**Review by andydotxyz — CHANGES_REQUESTED** [review]
Merge looks good.
However since this was opened we added the "Since:" tag - if you could add that to any new public APIs (I think it's just one) then we're good to go

File: test/test.go (hunk 8 above)
Comment (toaster) [reply test/test.go:45]: And If you pick this for a 1.4.x or a 1.5.0 you’ll adjust it?

Comment (andydotxyz) [reply test/test.go:47]: Sure thing :)

**Review by andydotxyz — APPROVED** [review]
